File: app/src/main/java/com/persona5dex/searchSuggestions/PersonaNameProvider.java
package com.persona5dex.searchSuggestions;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.database.Cursor;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.persona5dex.dagger.contentProvider.ContentProviderContextModule;
import com.persona5dex.dagger.contentProvider.DaggerContentProviderComponent;
import com.persona5dex.models.room.SearchSuggestionDao;

import java.util.List;

import javax.inject.Inject;

/**
 * Created by Rechee on 7/3/2017.
 */

public class PersonaNameProvider extends ContentProvider {

    public static final String FUSION_SUGGEST_PATH = "fusion";
    private SearchSuggestionDao dao;

    @Inject
    SearchSuggestionCursorProvider searchSuggestionCursorProvider;

    @Override
    public boolean onCreate() {
        DaggerContentProviderComponent
                .builder()
                .contentProviderContextModule(new ContentProviderContextModule(getContext()))
                .build()
                .inject(this);
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        final List<String> pathSegments = uri.getPathSegments();
        String query = pathSegments.get(pathSegments.size() - 1).toLowerCase();
        String suggestPath = pathSegments.get(0);

        boolean onlyPersonas = suggestPath.equals(FUSION_SUGGEST_PATH);
        return searchSuggestionCursorProvider.getSearchSuggestionCursor(String.format("%%%s%%", query), onlyPersonas);
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        return null;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }
}

File: app/src/main/java/com/persona5dex/searchSuggestions/PersonaNameProvider.kt
package com.persona5dex.searchSuggestions

import android.content.ContentProvider
import android.content.ContentValues
import android.database.Cursor
import android.net.Uri
import com.persona5dex.dagger.contentProvider.ContentProviderComponent
import com.persona5dex.dagger.contentProvider.ContentProviderContextModule
import com.persona5dex.dagger.contentProvider.DaggerContentProviderComponent
import javax.inject.Inject

/**
 * Created by Rechee on 7/3/2017.
 */
class PersonaNameProvider : ContentProvider() {

    @Inject
    internal lateinit var searchSuggestionCursorProvider: SearchSuggestionCursorProvider

    private lateinit var component: ContentProviderComponent

    override fun onCreate(): Boolean {
        component = DaggerContentProviderComponent
                .builder()
                .contentProviderContextModule(ContentProviderContextModule(context))
                .build()
        component.inject(this)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        val pathSegments = uri.pathSegments
        val query = pathSegments[pathSegments.size - 1].toLowerCase()
        val suggestPath = pathSegments[0]
        val onlyPersonas = suggestPath == FUSION_SUGGEST_PATH
        return searchSuggestionCursorProvider.getSearchSuggestionCursor(String.format("%%%s%%", query), onlyPersonas, component.gameType())
    }

    override fun getType(uri: Uri): String? = null
    override fun insert(uri: Uri, values: ContentValues?): Uri? = null
    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int = 0
    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int =
            0

    companion object {
        private const val FUSION_SUGGEST_PATH = "fusion"
    }
}